### PR TITLE
Fix xxe issues

### PR DESCRIPTION
**Pull request by ChamaraPhilipsuom**
Fixed security threats of Improper Restriction of XML External Entity Reference.

### CODE REVIEW

File: components/entitlement/org.wso2.carbon.identity.entitlement.common/src/main/java/org/wso2/carbon/identity/entitlement/common/util/CarbonEntityResolver.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package org.wso2.carbon.identity.entitlement.common.util;
```
Comment (pulasthi7) [inline components/entitlement/org.wso2.carbon.identity.entitlement.common/src/main/java/org/wso2/carbon/identity/entitlement/common/util/CarbonEntityResolver.java]: This should come below the licence

File: components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/PolicyEditorService.java [hunk 2]
```diff
@@ -173,6 +193,8 @@ public String formatXML(String xml) {
 
             xml = baos.toString("UTF-8");
 
+        } catch (ParserConfigurationException pce) {
+            throw new IllegalArgumentException("Failed to setup repository: ");
```
Comment (pulasthi7) [inline components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/PolicyEditorService.java]: exception object is not included in the wrapped exception. Provide some context information in the message.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/pap/PAPPolicyReader.java [hunk 3]
```diff
@@ -62,14 +69,22 @@
...
+
         // now use the factory to create the document builder
         try {
-            builder = factory.newDocumentBuilder();
+            documentBuilderFactory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+            builder = documentBuilderFactory.newDocumentBuilder();
+            builder.setEntityResolver(new CarbonEntityResolver());
             builder.setErrorHandler(this);
         } catch (ParserConfigurationException pce) {
             throw new IllegalArgumentException("Filed to setup repository: ");
```
Comment (pulasthi7) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/pap/PAPPolicyReader.java]: exception object is not included in the wrapped exception. Provide some context information in the message.

File: components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/util/CarbonEntityResolver.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package org.wso2.carbon.policyeditor.util;
```
Comment (pulasthi7) [inline components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/util/CarbonEntityResolver.java]: Move this below licence

File: components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/util/CarbonEntityResolver.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package org.wso2.carbon.identity.user.registration.util;
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/util/CarbonEntityResolver.java]: move this below licence header

File: components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/util/CarbonEntityResolver.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+public class CarbonEntityResolver implements EntityResolver {
+
+    @Override
+    public InputSource resolveEntity(String publicId, String systemId) throws SAXException,
+            IOException {
+        throw new SAXException("Possible XML External Entity (XXE) attack. Skip resolving entity");
+    }
+
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/util/CarbonEntityResolver.java]: remove the extra new lines

File: components/entitlement/org.wso2.carbon.identity.entitlement.common/src/main/java/org/wso2/carbon/identity/entitlement/common/util/CarbonEntityResolver.java (hunk 1 above)
Comment (ChamaraPhilipsuom) [inline components/entitlement/org.wso2.carbon.identity.entitlement.common/src/main/java/org/wso2/carbon/identity/entitlement/common/util/CarbonEntityResolver.java]: Fixed at #1682

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/pap/PAPPolicyReader.java (hunk 3 above)
Comment (ChamaraPhilipsuom) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/pap/PAPPolicyReader.java]: Fixed at #1682

File: components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/PolicyEditorService.java (hunk 2 above)
Comment (ChamaraPhilipsuom) [inline components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/PolicyEditorService.java]: Fixed at #1682

File: components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/util/CarbonEntityResolver.java (hunk 4 above)
Comment (ChamaraPhilipsuom) [inline components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/util/CarbonEntityResolver.java]: Fixed at #1682

File: components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/util/CarbonEntityResolver.java (hunk 5 above)
Comment (ChamaraPhilipsuom) [inline components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/util/CarbonEntityResolver.java]: Fixed at #1682

File: components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/util/CarbonEntityResolver.java (hunk 6 above)
Comment (ChamaraPhilipsuom) [inline components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/util/CarbonEntityResolver.java]: Fixed at #1682

File: components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/PolicyEditorService.java [hunk 7]
```diff
@@ -173,6 +193,8 @@ public String formatXML(String xml) {
...
         } catch (Exception e) {
             log.error(e);
```
Comment (mohanvive) [inline components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/PolicyEditorService.java]: Provide proper context message with stack trace.. Don't just log only the stack trace.

Comment (ChamaraPhilipsuom) [inline components/policy-editor/org.wso2.carbon.policyeditor/src/main/java/org/wso2/carbon/policyeditor/PolicyEditorService.java]: Fixed at #1682

File: components/entitlement/org.wso2.carbon.identity.entitlement.common/src/main/java/org/wso2/carbon/identity/entitlement/common/InMemoryPersistenceManager.java [hunk 8]
```diff
@@ -114,6 +123,26 @@ private PolicyEditorDataHolder buildDataHolder(String type, String xmlConfig) th
...
     }
 
+    /**
+     * * This method provides a secured document builder which will secure XXE attacks.
+     *
+     * @return DocumentBuilder
+     * @throws ParserConfigurationException
+     */
+    private DocumentBuilder getSecuredDocumentBuilder() throws ParserConfigurationException {
+
```
Comment (prabathabey) [inline components/entitlement/org.wso2.carbon.identity.entitlement.common/src/main/java/org/wso2/carbon/identity/entitlement/common/InMemoryPersistenceManager.java]: Remove unnecessary new line.

File: components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/UserRegistrationService.java [hunk 9]
```diff
@@ -383,4 +390,24 @@ private TenantRegistrationConfig getTenantSignUpConfig(int tenantId) throws Iden
...
     }
 
+    /**
+     * * This method provides a secured document builder which will secure XXE attacks.
+     *
+     * @return DocumentBuilder
+     * @throws ParserConfigurationException
+     */
+    private DocumentBuilder getSecuredDocumentBuilder() throws ParserConfigurationException {
+
```
Comment (prabathabey) [inline components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/UserRegistrationService.java]: Remove unnecessary new line.

File: components/entitlement/org.wso2.carbon.identity.entitlement.common/src/main/java/org/wso2/carbon/identity/entitlement/common/InMemoryPersistenceManager.java (hunk 8 above)
Comment (ChamaraPhilipsuom) [inline components/entitlement/org.wso2.carbon.identity.entitlement.common/src/main/java/org/wso2/carbon/identity/entitlement/common/InMemoryPersistenceManager.java]: Fixed at #1682

File: components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/UserRegistrationService.java (hunk 9 above)
Comment (ChamaraPhilipsuom) [inline components/user-mgt/org.wso2.carbon.identity.user.registration/src/main/java/org/wso2/carbon/identity/user/registration/UserRegistrationService.java]: Fixed at #1682

**Pull request merged**